Speed up find_cell searches for geometries with many z-planes

smharper opened the pull request:
This PR is aimed at speeding up multiphysics calculations that typically cut the geometry up with a bunch of z-planes.  Neighbor lists cannot be used when crossing lattice and universe boundaries which results in a very expensive search with finely discretized models.  This PR cuts down the cost of those `find_cell`s by performing a binary search on the z-planes.  After the binary search, the code can just check the small list of cells that can be found between the two bounding planes.

I did a timing test on my quarter-assembly multiphysics model with 37 axial regions.  When combined with #1208, I see a 60% increase in the neutrons per second.

There is also a ton of improvements we can make later to build off of this PR.
* Right now it only considers z-planes, but it is natural to extend the work to the other planes, cylinders, and spheres.
* This code can be further optimized by better handling cells that do not contain z-planes.  Currently those cells are added to all of the possible search outputs, e.g. the code doesn't check to see if a spherical cell is contained between a subset of the z-planes.
* In the future, we could also potentially have the code automatically detect universes with many cells (such as TRISO compacts) and add fake surfaces that aren't used in any region specifications but accelerate the searches.

paulromano commented on src/cell.cpp:
```diff
@@ -635,6 +635,93 @@ void DAGCell::to_hdf5(hid_t group_id) const { return; }
+    // Iterate over partitions, and add this cell to each appropriate one.
+    // Since surfs_ is sorted, we know this cell belongs to every partition
+    // between min_token and max_token.
+    bool in_partition = min_token < 0;
```
What happens if someone has defined a universe with lots of z-planes, but they don't define any cell that has `-min_zplane` (maybe because it gets cut off by virtue of being filled in another cell)? In that case, `min_token` would be positive, right?

smharper replied:
Yeah, we actually already have a case in the test suite that triggers this kind of weird condition.  The void regression test has enough z-planes to trigger the partitioning and it has a bunch of cells that are bound on one side by z-plane and by a sphere on the other.  In this case you can have a `min_token` that is negative.  (In this case, `min_token` is also equal to `max_token`.  So the code here assumes that the cell is present in the bottom partition and then every partition up to the `max_token == min_token` z-plane.)

In the normal case that we have a cell sandwiched between two z-planes then `min_token` is positive and `max_token` is negative.  I just now realized how counter-intuitive that is so maybe I should rename them to `lower_token` and `upper_token`

paulromano replied:
I'm not sure I fully understand yet, but it sounds like you've accounted for this case. Maybe after the renaming things will be more obvious to me :)

pshriwise commented:
@paulromano I was thinking the same thing regarding spatial trees. This is a great start and a nice way to take advantage of already existing planar divisions. This is a sort of 1-D kd-tree now and could certainly be extended to other types of divisions, as @smharper said. A (z,r) tree would be interesting for the types of geometries we see most often, but this would be tricky with cylinders.. 

Ideally, we can implement bounding boxes for all cells at some point. Then we can experiment with using a kd-tree or Octree. Hopefully that would set us up well to try out some new transport methods (thinking delta tracking here).

smharper commented on src/cell.cpp:
```diff
@@ -635,6 +635,93 @@ void DAGCell::to_hdf5(hid_t group_id) const { return; }
+    // Since surfs_ is sorted, we know this cell belongs to every partition
+    // between min_token and max_token.
+    bool in_partition = min_token < 0;
+    for (auto i = 0; i < surfs_.size(); ++i) {
```
Oh, whoops.  I was assuming `auto` would become a `size_t` in cases like this.  C++ can be tricky sometimes!

paulromano replied:
I know, I wish it did

smharper commented on src/cell.cpp:
```diff
@@ -635,6 +635,93 @@ void DAGCell::to_hdf5(hid_t group_id) const { return; }
+  for (auto i_cell : univ.cells_) {
+    // Find the tokens for bounding z-planes.
+    int32_t min_token = 0, max_token = 0;
+    double min_z, max_z;
```
I think we should make an exception to ES.10 for cases like this.  The ES.10 example has a bunch of arrays which is certainly confusing, but here the types are super simple.  I think adding a on of separate lines to declare blocks of simple variables like these hurts rather than helps readability.

paulromano replied:
Yeah, I don't feel too strongly about it for this case, so I'm fine leaving as is

smharper commented:
In addition to addressing your comments, I've tried reworking the logic in the `UniversePartitioner` constructor so that it is hopefully more clear but equivalent.  I also moved `get_cells` into the .cpp file, and it looks like the runtime impact is pretty small.

paulromano commented on src/cell.cpp:
```diff
@@ -635,6 +635,143 @@ void DAGCell::to_hdf5(hid_t group_id) const { return; }
+
+  // Define a lambda for comparing z-planes by their location, and use it to
+  // sort the surfs_ member.
+  std::sort(surfs_.begin(), surfs_.end(),
```
Thinking about this a little more, it would be much easier to just use `std::set` instead of `std::unordered_set` above and then just construct `surfs_` by iterating over the `set`.

paulromano commented on src/cell.cpp:
```diff
@@ -635,6 +635,143 @@ void DAGCell::to_hdf5(hid_t group_id) const { return; }
+          }
+        }
+      }
+    }
```
This block of code could be combined with the first block where you are building `surf_set` since it's all the same up until the point you're comparing to `min_z`.

paulromano replied:
Nevermind, I take that back. I see this is specific for each cell.

paulromano commented on src/cell.cpp:
```diff
@@ -635,6 +635,143 @@ void DAGCell::to_hdf5(hid_t group_id) const { return; }
+    }
+
+    // Add the cell to all relevant partitions.
+    for (int i = first_partition; i < last_partition + 1; ++i) {
```
I personally like `i <= last_partition`

paulromano commented:
@smharper The new logic for building the partitions is indeed much clearer. I was able to wrap my brain around it this time. Thanks! This is good to go after the last few comments are resolved.
